### ISSUE TITLE
Custom gradient tape-based fit() function is slower than Tf2 fit() and has memory leaking

### ISSUE THREAD

**Issue by Alf-Z** [tensorflow/tensorflow#48732]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04.1 LTS
- TensorFlow version (use command below): v2.3.0-54-gfcc4b966f1 2.3.1
- Python version: Python 3.8.5
- CUDA/cuDNN version: CUDA Version 11.0.207
- GPU model and memory: NVIDIA Quadro RTX 5000, 16 GB GDDR6

I wrote a custom `fit()` function using gradient tape in TF2. The `@tf.function` decorator has been used in a few specific locations to disable eager execution, which is supposed to speed up the whole code.

My custom fit has three main issues:

1-It's a lot slower than its TF2 counterpart (decorator `@tf.function` on the `train_batch` function does not improve the speed, while the decorator placed above the `fit()` function does). The time indicated on the output (see image attached below) of the custom `fit()` is wrong for some reason, in reality it is constant, around 10 seconds per each epoch.

2-Memory usage does increase linearly (see plot at the bottom).

3-Code gets killed even before reaching the maximum amount of GPU memory usage (see code output at the bottom).

Below you can find the complete code, you should be able to run it directly, as it does not require any data or any additional line of code.
```
import tensorflow as tf
import tensorflow.keras as keras
from tensorflow.keras.layers import Input, Dense, Dropout, Conv1D, Flatten, MaxPooling1D
from tensorflow.keras.models import Model
import numpy as np
import warnings
import psutil
import time
import matplotlib.pyplot as plt


### DEFINE NEURAL NETWORK
def baseline_network():
    input = Input(shape=(128, 11))
    layer = Conv1D(64, 5, activation='relu', padding='same')(input) 
    layer = Flatten()(layer)
    layer = Dense(1024, activation='relu')(layer)
    output = Dense(257, activation=None)(layer)
                  
    model = Model(inputs=input, outputs=output)
    model.compile(optimizer='adam',
                  loss='mse', 
                  metrics=['mse'])
   
    return model


### DATA GENERATOR
def generator():
    
    # initialize numpy tensors
    X_out = np.zeros([6000, 128, 11])
    Y_out = np.zeros([6000, 257])

    ### CREATE DATA BATCHES ###
    while True:
        for i in range(32):
        
            yield X_out, Y_out


### CUSTOM FIT USING GRADIENT TAPE
class runGradientTape:
    
    def __init__(self, model, cost_function, iterator_train, steps_train, max_epochs, batch_size):
        
        self.model = model
        self.cost_function = cost_function
        self.iterator_train = iterator_train
        self.steps_train = steps_train
        self.max_epochs = max_epochs
        self.batch_size = batch_size


    def lossWrapper(self):
        
        #@tf.function
        def lossFunction(y_true, y_pred):
            # calculating loss and squeezing single dimensions away
            loss = tf.squeeze(self.cost_function(y_pred, y_true))
            # calculate mean over batches
            loss = tf.reduce_mean(loss)
            # return the loss
            return loss
        # returning the loss function as handle
        return lossFunction
    
    
    @tf.function
    def train_batch(self, x_batch_train, y_batch_train):
        # run gradient taping
        with tf.GradientTape() as tape:
            y_hat = self.model(x_batch_train, training=True)
            loss_value = self.loss_fn(y_batch_train, y_hat)

        # calculate gradients
        grads = tape.gradient(loss_value, self.model.trainable_variables)
        self.model.optimizer.apply_gradients(zip(grads, self.model.trainable_variables)) 
        
        return loss_value
    
    
    @tf.function
    def fit(self):
        
        # others
        warnings.filterwarnings("ignore")

        self.loss_fn = self.lossWrapper()
        progbar = tf.keras.utils.Progbar(self.steps_train)
        
        memory_used = []
        
        # ITERATE OVER EPOCHS 
        for epoch in range(self.max_epochs):
            
            print('Epoch = %d/%d' %(epoch+1, self.max_epochs))
            
            # iterate over steps in training generator
            for i in range(self.steps_train):
                
                x_batch_train, y_batch_train = next(self.iterator_train)
                
                loss_value = self.train_batch(x_batch_train, y_batch_train)
                progbar.update(i+1)
                
                memory_used.append(psutil.virtual_memory().used / 2 ** 30)
                #print('   memory used: ', memory_used[-1])
        
        #plt.plot(memory_used)
        #plt.title('Memory usage vs batch')
        #plt.savefig('mem_usage')

                
# parameters
steps_train = 100       
n_epochs = 5
batch_size = 32
baseline =  baseline_network()   
iterator_train = generator()

# TF2 fit()
print("TF2 fit")
start_time = time.time()
history = baseline.fit(iterator_train, steps_per_epoch=steps_train, epochs=n_epochs)
stop_time = time.time()
print("time elapsed (TF2 fit): ", stop_time-start_time)

#reset parameters
del baseline, iterator_train
baseline =  baseline_network()   
iterator_train = generator()

# custom fit
print("Custom fit")
start_time = time.time()
run_gradient_tape = runGradientTape(baseline, keras.losses.MSE, iterator_train, steps_train, n_epochs, batch_size)  
run_gradient_tape.fit()
stop_time = time.time()
print("time elapsed (custom fit): ", stop_time-start_time)
```

Some of the suggestions I tried so far:

1- `gs.collect()` does not give any improvement in this case.

2- profiling the code using tensorboard (check profile and memory usage below).

Could you please give me any suggestion on how to solve the three previously mentioned problems?

Code output
![](https://i.stack.imgur.com/LEZkN.png)

Memory usage
![](https://i.stack.imgur.com/IA4wr.png)
![](https://i.stack.imgur.com/Oj8Tc.png)

Timing:
![](https://i.stack.imgur.com/OpaD3.png)

**Comment by tilakrayal** [comment]
@ymodak

I have tried in colab with TF version 2.4, 2.5.orc1 and nightly version and noticed that session is being crashed. Please, find the [gist](https://colab.research.google.com/gist/tilakrayal/8d0d45222b42d0f9ae4b7af8a412affc/48732.ipynb) here. Thanks!

**Comment by Alf-Z** [comment]
> 
> 
> @ymodak
> 
> I have tried in colab with TF version 2.4, 2.5.orc1 and nightly version and noticed that session is being crashed. Please, find the [gist](https://colab.research.google.com/gist/tilakrayal/8d0d45222b42d0f9ae4b7af8a412affc/48732.ipynb) here. Thanks!

Hi, I believe it has something to do with the RAM available in Colab. I do not use this service but, as far as I understood, maximum RAM should be around 11-12 GB, please correct me if I am wrong. 
The code crashes around batch 121-143. If you see the linear memory usage plot, we should be around that value.
However, my code does not crash on my GPU, but it gets killed at the very end, after the very last batch is fed. You may want to just reduce the number of epochs to 1, and see if this problem still occurs in Colab.

**Comment by tomytsai1** [comment]
Tried to reproduce this and saw that @tf.function is not marked correctly:
Because @tf.function's first execution is expensive (which is used to do [function tracing](https://www.tensorflow.org/guide/intro_to_graphs#when_is_a_function_tracing)), marking the custom loop function as a @tf.function will cause the whole training process to be under tracing, which is slow and uses extra mem, eventually killing the training process when the memory are used up.

The convention is to mark the single training step as a @tf.function, so only the first training batch will have the graph tracing overhead (if the data type and shape remain the same for the remaining data). 

Please see [this colab](https://colab.research.google.com/gist/tilakrayal/8d0d45222b42d0f9ae4b7af8a412affc/48732.ipynb#scrollTo=FhiHyJoRLmoe) as an example. We see that the custom fit runs about the same as (or slightly faster than) the tf2 fit, and no apparent memory increase after the custom training.
